File: data_structures/queue/queue.go
package queue

import (
  "fmt"
)

type queueNode struct {
  data int
  min int
  max int
  next *queueNode
}

type Queue struct {
  head *queueNode
  tail *queueNode
}

func NewQueue() *Queue {
  return &Queue{}
}

func (q *Queue) Queue(value int) error {
  node := &queueNode{
    data: value,
    min: value,
    max: value,
  }

  if q.head == nil {
    q.head = node

    if q.tail == nil {
      q.tail = q.head
    }

    return nil
  }

  if value > q.head.min {
    node.min = q.head.min
  }

  if value < q.head.max {
    node.max = q.head.max
  }

  tmp := q.head
  tmp.next = node
  q.head = node

  if q.tail == nil {
    q.tail = q.head
  }

  return nil
}

func (q *Queue) Dequeue() (int, error) {
  if q.tail == nil {
    return -1, fmt.Errorf("can't dequeue from an empty queue")
  }

  tmp := q.tail
  q.tail = tmp.next
  return tmp.data, nil
}

func (q *Queue) Min() (int, error) {
  if q.head == nil {
    return -1, fmt.Errorf("can't get the min value of an empty queue")
  }

  return q.head.min, nil
}

func (q *Queue) Max() (int, error) {
  if q.head == nil {
    return -1, fmt.Errorf("can't get the max value of an empty queue")
  }

  return q.head.max, nil
}

File: data_structures/queue/queue.js
class QueueNode {
  constructor(value) {
    this.value = value;
    this.min = value;
    this.max = value;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  queue(value) {
    let node = new QueueNode(value);

    if (this.head == null) {
      this.head = node;

      if (this.tail == null) {
        this.tail = this.head;
      }

      return;
    }

    node.min = value < this.head.min ? value : this.head.min;
    node.max = value > this.head.max ? value : this.head.max;

    let tmp = this.head;
    tmp.next = node;
    this.head = node;

    if (this.tail == null) {
      this.tail = this.head;
    }
  }

  dequeue() {
    if (this.tail == null) {
      return null;
    }

    let tmp = this.tail;
    this.tail = tmp.next;
    return tmp.value;
  }

  min() {
    return this.head ? this.head.min : -1;
  }

  max() {
    return this.head ? this.head.max : -1;
  }
}

var queue = new Queue();
queue.queue(1);
queue.queue(3);
queue.queue(2);
console.log(queue.min());
console.log(queue.max());
console.log('dequeue!', queue.dequeue());
console.log(queue.min());
console.log(queue.max());
console.log('dequeue!', queue.dequeue());
console.log(queue.min());
console.log(queue.max());
console.log('dequeue!', queue.dequeue());
